File: adl/devices/digital_output.py
import logging

from collections import namedtuple

from yapsy.IPlugin import IPlugin

from adl.devices.generic_device import GenericDevice

class DigitalOutput(GenericDevice, namedtuple("DigitalOutput", ["name", "pin"])):

	__slots__ = ()

	@property
	def setup(self):
		return "{name}.setup();".format(name=self.cname())

	@property
	def command_handler(self):
		return "return {name}.command_handler(command, reply);".format(name=self.cname())

	@property
	def sources(self):
		return ["digital-output.cpp"]

	@property
	def includes(self):
		return ["digital-output.h"]

	@property
	def declarations(self):
		return "static DigitalOutput {name} = DigitalOutput({pin});".format(
			name=self.cname(), pin=self.pin.number)

class DigitalOutputPlugin(IPlugin):
	def activate(self):
		pass

	def deactivate(self):
		pass

	def get(self, device):
		return DigitalOutput(device.name, device.pins[0])

	def set_log_level(self, level):
		logging.getLogger(__name__).setLevel(level)

File: adl/devices/digital-output.cpp
#include <Arduino.h>

#include "device.h"
#include "adl.h"

#include "digital-output.h"

DigitalOutput::DigitalOutput(int pin)
{
	m_pin = pin;
}

void DigitalOutput::reset()
{
	digitalWrite(m_pin, LOW);
}

void DigitalOutput::setup()
{
	pinMode(m_pin, OUTPUT);
	this->reset();
}

int DigitalOutput::command_handler(char const * const command, char * reply)
{
	int reply_length = 0;
	if (command[0] == '1')
	{
		digitalWrite(m_pin, HIGH);
		strcpy(reply, "OK");
		reply_length = 2;
	}
	else if (command[0] == '0')
	{
		digitalWrite(m_pin, LOW);
		strcpy(reply, "OK");
		reply_length = 2;
	}
	else
	{
		reply[0] = '?';
		reply_length = 1;
	}

	return reply_length;
}
